source/common/stats/thread_local_store.cc

```diff
@@ -219,6 +219,8 @@ StatType& ThreadLocalStoreImpl::ScopeImpl::safeMakeStat(
         make_stat(parent_.alloc_, truncated_name, std::move(tag_extracted_name), std::move(tags));
     if (stat == nullptr) {
       parent_.num_last_resort_stats_.inc();
+      std::vector<Tag> tags;
+      std::string tag_extracted_name = parent_.getTagsForName(name, tags);
```

Comment:
If we didn't make the stat above do we need to do this again? I think the code before might have been safe? @jmarantz?

Reply:
if `make_stat` return `nullptr`, the `tag_extracted_name`  and `tags` was already move, 
When you execute make_stat again, both variables are already null.

Reply:
Interesting. If so, can we add a test case to catch this?

Reply:
Testing code:
```
#include <iostream>
#include <string>

void ignoreString(std::string&& str) {}

int main() {
  std::string hello = "Hello, world!";
  ignoreString(std::move(hello));
  std::cout << "hello='" << hello << "'" << std::endl;
  return 0;
}
```
and running it:
```
g++ move_string.cc;./a.out
hello='Hello, world!'
```

The code as is is a lot like the sample above. It works because when makeStat fails it doesn't capture the string in a destination, so the move() never finishes and is still available to move to a new destination. I realize this is somewhat counter-intuitive, and I have a refactor pending to clean this up anyway.

Reply:
```
#include <iostream>
#include <string>

void ignoreString(std::string&& str) {
  std::string txt = std::move(str);
}

int main() {
  std::string hello = "Hello, world!";
  ignoreString(std::move(hello));
  std::cout << "hello='" << hello << "'" << " size: " << hello.size() << std::endl;
  return 0;
}
hello='' size: 0
```

I guess if you don't do anything, it will be optimized by the compiler.

Reply:
I'm not sure what you mean by 'optimized by the compiler'.   A compiler optimization should not dictate whether 'hello' winds up empty or not.

Reply:
My understanding should be wrong,` std::string&&` This is an rvalue reference, so there is no actual move, but we can't guarantee that the `ignoreString` will not do the move inside.

Reply:
In the stats code, makeStats() does not move the tag_extracted_name when it fails, so you can try again with a second allocator.

Reply:
https://en.cppreference.com/w/cpp/utility/move uses the language: 
```
std::move is used to indicate that an object t may be "moved from", i.e.
allowing the efficient transfer of resources from t to another object.

In particular, std::move produces an xvalue expression that identifies
its argument t. It is exactly equivalent to a static_cast to an rvalue
reference type.
```

IMO std::move is not a very intuitive name; it's would have been better to call in std::rvalue_cast or similar.

Reply:
Yes, there really is no actual move, then I add a TODO, waiting for your refactor it, increase the readability of this part of the code?

Reply:
sure, if you like; that will just cause a conflict but I can resolve :)